### ISSUE TITLE
streams: make stream access thread-safe

### ISSUE THREAD

**Issue by dkochmanski** [McCLIM/McCLIM#585]
Right now when we write asynchronously to a stream output records get corrupted. Issue is reported here:

https://github.com/McCLIM/McCLIM/issues/315

This is opened as a separate task, because it is more an architecture enhancement. I plan to assign $600 bounty on this task before the next release, but if fixed earlier it will be paid anyway.

**Comment by dkochmanski** [comment]
https://www.bountysource.com/issues/65501924-streams-make-seos-access-thread-safe

**Comment by g--o** [comment]
> https://www.bountysource.com/issues/65501924-streams-make-seos-access-thread-safe

Hey is this still relevant?

**Comment by dkochmanski** [comment]
yes it is. there is even a preferred way how to fix it (unless you have another proposal which we could discuss): check if the stream thread is the same from which operation is invoked and if not queue an event with the operation (and define handle-event method which works on such events). I've tried first with locks, but that didn't work for various reasons (and had a non-negligible performance penalty).

**Comment by g--o** [comment]
@dkochmanski 
Hey, so I've been investigating for a while and with the documentation as bad as it is, it's pretty hard to figure out the exact source of the race condition and the flow of your architecture. I've seen your commits and the locking seem to be too general, and thus create deadlock in some of my tests. Note that even with a queue you'd need a lock and to know when exactly to dequeue, so understanding the race condition is crucial for solving the problem anyway.

As of my own findings - while i suspect as well that something's going on with recordings and repaints it definitely all starts with the the stream invoking scrolling methods like move-pane (according to my tests). Still, any pointers would be really great as to what your investigation led up to so far, so I can maybe try and continue from there to speed up the process. In particular: is the stream we interact with and the record stream the same? because this might be problematic as well

I'd really appreciate your thoughts on this, thanks!

**Comment by dkochmanski** [comment]
Event queue already is thread-safe (see file input.lisp). Output stream may record its output (when it is an output-recording-stream), so it is indeed a case that the stream we interact with is recording its output. That's basically when the corruption happens if we call it asynchronously. Here is where the race condition may come from:

1. CLIM loop receives a repaint event and sets recording to NIL on the stream
2. We call draw-rectangle* on said stream what sets recording to T
3. repaint carries on and repaint is recorded (changing i.e current-output-record)
4. everything is jumbled up from this point on

I've tried both approaches: with fine-grained locks (and that lead to a deadlock) and a  global lock (which is not reasonable). That lead me to a conclusion, that locking streams is not the right thing to do. Given that event-queue is already thread-safe and CLIM operates in its own loop, we could simply queue operations which need to be done, that will avoid the deadlocks and corruption.

Here comes a pseudocode of what I mean:
```
(defclass thread-safe-stream-mixin ()
  ((thread :initarg :thread)))

(defmethod draw-rectangle* :around ((s thread-safe-stream-mixin) …)
  (if (eql (current-thread) (stream-thread s))
    (call-next-method)
    (queue-event (make-instance 'lambda-event :fun (lambda () (draw-rectangle* s …)))))

;; now from the clim loop, when event is read
(defmethod handle-event ((stream thread-safe-stream-mixin) (event lambda-event))
  (funcall fun))
```

handle-event will be called in the loop, so we just schedule asynchronous calls for processing.

**Comment by g--o** [comment]
Oh so you're basically saying that repaint isn't called unless we scroll + my hypothesis that the recorded stream = interacted stream. That actually explains a lot of things, thanks!

**Comment by g--o** [comment]
@dkochmanski Hey so I've implemented this thing but it doesn't seem to work? The event mechanism itself is ok, as it works when set on all mediums (though obvs isn't what we want).

And I mean it isn't even triggered. When you say that draw-rectangle* is called with stream do you mean literally with stream as the first argument? Or do you mean the cursor? Because letting output-stream inherit does not work. 
In anycase the first argument of draw-rectangle (supposedly a sheet) is abused and really confusing...

Any ideas?

**Comment by g--o** [comment]
@dkochmanski hey so i tried queue-ing draw-rectangle* for output-stream, output-record, basic-pane, basic-medium to no avail, I tried move-sheet as well, and tried both queued to no avail.
(note: I didn't check thread though, just queued everything)

So either I missed the class that needs queuing or our method isn't gonna work...
thoughts? I'm pretty clueless at this point

**Comment by clynbech** [comment]
Just to be sure I understand the targets that needs fixing.

In the pseudo code, it says `defmethod draw-rectangle*` but I can only find a generic that is called `medium-draw-rectangle*`, is that the correct function we are talking about?

The explanation also says that the fix should apply to all graphic ops and textual output.

The graphic ops I assume are the following:
- medium-draw-point*
- medium-draw-points*
- medium-draw-line*
- medium-draw-lines*
- medium-draw-polygon*
- medium-draw-rectangle*
- medium-draw-ellipse*
- medium-draw-circle*
- medium-draw-text*

What about "textual output", what functions are related to that?

**Comment by dkochmanski** [comment]
hey o/

`draw-rectangle*`, (and `draw-ellipse*` etc) are all trampolines to `medium-draw-whatever*`. From the user perspective it is the trampoline what needs to be thread safe. Pseudo-code is of course wrong with this regard and I've meant `medium-draw-rectangle*` there indeed.

All high-level functions (like `draw-rectangle*` are calling `with-medium-options before calling `medium-draw-*`, so this macro may be the right place to queue the lambda event (it is important to honor what user specified, i.e if they call `(draw-rectangle* stream 0 0 10 10 :ink +red+)` then we want to preserve this ink (which is part of the medium state which is set by a with-medium-options macro).

Macro `with-medium-options` calls a generic (internal) function `do-graphics-with-options`, so it might be that the lambda-event should be queued from this method. Then we wouldn't need to track each individual medium function.

Regarding textual-output: in principle there are two ways to work with text: `draw-text*` function and the stream interfaces. The latter is a little more troublesome, because `format` may butcher strings to multiple stream-write-string calls, so you have no guarantee about number of calls to write a single string. If the programmer wraps their call to format in (with-medium-options …) given you take this approach for drawing operations, then they are safe.

Now there is a question, what to do with asynchronous calls to stream-write-string? What we want to avoid at all cost is a corruption of the cursor position (that part was tackled by @g--o), but there are other gotchas. Imagine scenario with two threads where one draws rectangles with a random ink and other calls format on a stream. First thread changes ink, and that influences the medium state (i.e its ink), so the text also is drawn with a varying ink.

I'm inclined to believe that the best solution from such scenarios is to demand calling stream text operations from inside with-medium-options and if (i.e) stream-write-string is called in a thread which is not local for the stream, then we should signal an error with a meaningful message, that operation should be called either inside the stream thread or wrapped in `with-medium-options`.

As you may tell I'm not certain about the right solution, I've spent some time exploring lock-based approach and gained some intuition with it, but that's all. If you can propose something more robust or appealing in other way please do and we may discuss it. Note, that we other devs hang out on #clim channel on freenode, so if you are looking for a productive discussion you may try to join (keep in mind that people may not be around the exact moment you join).

If you have other questions please ask them, I'll try to answer them best I can.

**Comment by dkochmanski** [comment]
for the future (you may already know this), when you leave a question on IRC and leave, you may check if there is an answer in our log. I've answered your question about events here: https://irclog.tymoon.eu/freenode/%23clim?around=1568727627#1568727627

**Comment by clynbech** [comment]
Sorry, it was not to be impatient and it was not a very important
question, just a small reflection.

I was sure I could pick it up, although I did not know in advance how,
so I am thankfull for the pointer.

/Christian


-----------------------
On Tue, Sep 17 2019, Daniel Kochmanski wrote:

for the future (you may already know this), when you leave a question on IRC and leave, you may check if there
is an answer in our log. I've answered your question about events here:
https://irclog.tymoon.eu/freenode/%23clim?around=1568727627#1568727627

**Comment by goose121** [comment]
It seems like some other things that would need to be looked at are `stream-current-output-record`, `stream-current-text-output-record`, `stream-recording-p`, and `stream-drawing-p`; if two threads are simultaneously trying to use the same stream (even if they are doing so via the event queue), they may end up clobbering each other's values of these slots when trying to do complex output, so thread-local values for these slots may have to be maintained somehow.
